FerramAerospaceResearch.Base/UnityJobs/OcclusionJobs.cs

```diff
@@ -0,0 +1,381 @@
+    {
+        [ReadOnly] public NativeArray<quaternion> quaternions;
+        [ReadOnly] public NativeArray<float> distances;
+        [ReadOnly] public NativeArray<float> cutoffDistance;
+        public NativeList<SphereDistanceInfo> sdiList;
+        [WriteOnly] public NativeList<quaternion> orderedQuaternions;
+
+        public void Execute()
+        {
+            for (int i = 0; i < quaternions.Length; i++)
+            {
+                if (distances[i] < cutoffDistance[0])
+                {
+                    SphereDistanceInfo sdi = new SphereDistanceInfo
+                    {
+                        q = quaternions[i],
+                        distance = distances[i]
+                    };
+                    sdiList.Add(sdi);
+                }
+            }
+            NativeSortExtension.Sort(sdiList);
+            for (int i = 0; i < sdiList.Length; i++)
+                orderedQuaternions.Add(sdiList[i].q);
+        }
+    }
+
+    [BurstCompile]
+    public struct MakeQuaternionIndexMapJob : IJobParallelFor
+    {
+        [ReadOnly] public NativeArray<quaternion> arr;
+        [WriteOnly] public NativeHashMap<quaternion, int>.ParallelWriter map;
+
+        public void Execute(int index)
+        {
+            map.TryAdd(arr[index], index);
+        }
+    }
+
+    // Given anl array of items and a map, update the map s.t. map[item] = min(current value, index in array)
+    // Can't parallelize the read/write cycle of the NativeHashMap, so can't do .ParallelWriter
+    [BurstCompile]
+    public struct ListToIndexedMap : IJob
+    {
+        [ReadOnly] public NativeArray<quaternion> sorted;
+        public NativeHashMap<quaternion, int> map;
+        public void Execute()
+        {
+            for (int index=0; index < sorted.Length; index++)
+            {
+                quaternion q = sorted[index];
+                if (map.TryGetValue(q, out int i) && index < i)
+                {
+                    map[q] = index;
+                }
+                else
+                    map.TryAdd(q, index);
+            }
+        }
+    }
+
+    [BurstCompile]
+    public struct BucketSortByPriority : IJobParallelFor
+    {
+        [ReadOnly] public NativeArray<quaternion> quaternions;
+        [ReadOnly] public NativeHashMap<quaternion, int> priorityMap;
+        [ReadOnly] public NativeHashMap<quaternion, EMPTY_STRUCT> completed;
+        [ReadOnly] public int maxIndexForPriority0;
+        [WriteOnly] public NativeMultiHashMap<int, int>.ParallelWriter map;
+
+        public void Execute(int index)
+        {
+            if (!completed.ContainsKey(quaternions[index]))
+            {
+                int pri = 2;
+                if (priorityMap.TryGetValue(quaternions[index], out int i))
+                    pri = math.min(pri, (i <= maxIndexForPriority0) ? 0 : 1);
+                map.Add(pri, index);
+            }
+        }
+    }
+
+    [BurstCompile]
+    public struct ClearNativeHashMapIntFloat : IJob
```

Comment:
Will making the job generic not compile? It would avoid using types in job names
https://docs.unity3d.com/Packages/com.unity.burst@1.5/manual/docs/OptimizationGuidelines.html#generic-jobs

Reply:
Implemented in 53809f6